node/src/store.rs

```diff
@@ -1,14 +1,326 @@
+        let genesis_height = genesis.height().value();
+
+        InMemoryStore {
+            headers: DashMap::from_iter([(genesis_hash, genesis)]),
+            height_to_hash: DashMap::from_iter([(genesis_height, genesis_hash)]),
+            head_height: AtomicU64::new(genesis_height),
+        }
+    }
+
+    pub fn get_head_height(&self) -> u64 {
+        self.head_height.load(Ordering::Acquire)
+    }
+
+    #[instrument(err)]
+    pub fn append_continuous(&self, header: ExtendedHeader) -> Result<(), StoreError> {
+        let hash = header.hash();
+        let height = header.height();
+
+        // lock both maps to ensure consistency
+        // this shouldn't deadlock as long as we don't hold references across awaits if any
+        // https://github.com/xacrimon/dashmap/issues/233
+        let hash_entry = self.headers.entry(hash);
+        let height_entry = self.height_to_hash.entry(height.into());
+
+        if matches!(hash_entry, Entry::Occupied(_)) {
+            return Err(StoreError::HashExists(hash));
+        }
+        if matches!(height_entry, Entry::Occupied(_)) {
+            return Err(StoreError::HeightExists(height.into()));
+        }
+        if self.get_head_height() + 1 != height.value() {
+            return Err(StoreError::NonContinuousAppend(
+                self.get_head_height(),
+                height.value(),
+            ));
+        }
+
+        info!("Will insert {hash} at {height}");
+        hash_entry.insert(header);
+        height_entry.insert(hash);
+
+        self.head_height.store(height.value(), Ordering::Release);
+
+        Ok(())
+    }
+
+    #[instrument(err)]
+    pub fn get_head(&self) -> Result<ExtendedHeader, StoreError> {
+        let head_height = self.head_height.load(Ordering::Acquire);
+        if head_height == 0 {
+            return Err(StoreError::NotFound);
+        }
+
+        let Some(head_hash) = self.height_to_hash.get(&head_height).as_deref().copied() else {
+            return Err(StoreError::LostStoreHead);
+        };
+
+        self.headers
+            .get(&head_hash)
+            .as_deref()
+            .cloned()
+            .ok_or(StoreError::LostHash(head_hash))
+    }
+
+    pub fn exists_by_hash(&self, hash: &Hash) -> bool {
+        self.headers.get(hash).is_some()
+    }
+
+    #[instrument(err)]
+    pub fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader, StoreError> {
+        self.headers
+            .get(hash)
+            .as_deref()
+            .cloned()
+            .ok_or(StoreError::NotFound)
+    }
+
+    pub fn exists_by_height(&self, height: u64) -> bool {
+        let Some(hash) = self.height_to_hash.get(&height).as_deref().copied() else {
+            return false;
+        };
+
+        self.headers.get(&hash).is_some()
+    }
+
+    #[instrument(err)]
+    pub fn get_by_height(&self, height: u64) -> Result<ExtendedHeader, StoreError> {
+        let Some(hash) = self.height_to_hash.get(&height).as_deref().copied() else {
```

Comment:
Small optimization, after you apply the suggestion on `exists_by_height` :

```suggestion
        if !exists_by_height(height) {
                return Err(StoreError::NotFound);        
        }

        let Some(hash) = self.height_to_hash.get(&height).as_deref().copied() else {
```